fetch.bs

```diff
@@ -4457,6 +4586,14 @@ steps:
 
  <li><p><a>Upgrade <var>request</var> to a potentially trustworthy URL, if appropriate</a>.
 
+ <li>
+  <p>Optionally, run <a>upgrade an HTTP request</a> algorithm on <var>request</var>.
+
+  <p class=note>HTTPS upgrading only applies to requests with <a>HTTP(S) scheme</a>s, but it's done
+  in <a>main fetch</a> instead of <a>HTTP fetch</a> to ensure that
+  <a>upgrade a mixed content <var>request</var> to a potentially trustworthy URL, if appropriate</a>
+  step runs next and applies to the upgraded request.
+
```

Comment:
Thank you! That does appear to be correct and does seem like something we ought to address (and test!).

Also reading this again the rationale here is flawed as HTTPS upgrading only applies when the destination is "document" whereas Mixed Content upgrading only applies for media destinations. So that needs fixing as well.

Reply:
Apologies for the delay here.

If I'm understanding this correctly, having the referrer determined after the upgrade sounds like the correct behavior: If there was a successful HTTPS Upgrade, we don't want to drop the referrer even if the original link was http. So the fix needs to be made to the HSTS step, and not here, is that correct?

> Also reading this again the rationale here is flawed as HTTPS upgrading only applies when the destination is "document" whereas Mixed Content upgrading only applies for media destinations. So that needs fixing as well.

Another reason to have the upgrade called in main fetch was to keep it close to all other automatic upgrades (mixed content, hsts, UIR). Given that HTTPS upgrades is different than all of these because it only applies to document navigations, is the implication that we should also move it to network-or-cache-fetch like the fallback? Or can we just remove the justification and keep it here?

Reply:
1. I tend to agree that it would make sense to determine the referrer based on the updated URL. I think that will also do the correct thing with regards to an upgrade followed by a downgrade.
   1. We should make sure this is tested.
1. There should be a follow-up issue to determine the interaction of referrer and HSTS. We should probably change it, but it would be clearest if done in a separate PR. Ideally with its own tests.
1. I think we should remove the note, but indeed leave the upgrade happening right where it is now. The rationale is that you want to make policy decisions early as everything else is impacted by them, but I'm not sure that stating that for each policy decision is worthwhile. Just something to keep in mind.
1. I will be taking an end-of-the-year break pretty soon so it's likely this won't be able to land until mid-January. Hopefully the above is enough to make meaningful forward progress until then.

Reply:
Thanks!

1. I'm adding a WPT for the HTTPS Upgrades & referrer interaction at https://github.com/web-platform-tests/wpt/pull/43676

2. I opened issue 1727 for the HSTS & referrer interaction. Since it's a preexisting issue, I'm not sure if it should be a blocker for this PR though.

3. I removed the note about mixed content autoupgrading from the upgrade step.

4. Happy holidays!